GeoJSON Code Review

lchans opened the pull request.

Commit:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

Added KML Coordinate class and implementation

juliawong committed:
refactored variable names

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

Commit:
Added some tests for coordinate class, changed the comments so that they are more relevant to the method

juliawong committed:
Added new GeoJSON file for demo

juliawong committed:
Redid the checking if feature collection and feature. Implemented the functions to parse the GeoJSON file and each GeoJSON object in the file.

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added GeoJSON file containing a large amount of markers around Spain

juliawong committed:
Implemented methods which created multiple map objects - toMarkers, toPolylines, toPolygons.
Added modified GeoJSON demo.

juliawong committed:
Added classes to create properties for each type of geometry. Updated comments in all files.

Commit:
Changed variable assignment, now hashmap

juliawong committed:
Updated comment for class

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Allowed for Properties classes to parse the coordinates

juliawong committed:
Refactored code

juliawong committed:
Added tests that test for the initialisation of a new Property object

Commit:
Classes now take care of reading in input and assigning them to appropriate values

juliawong committed:
Added some more comments and redid some. Ran the reformatter on the code in the kml directory

juliawong committed:
Readded comments documenting what is being thrown

juliawong committed:
Renamed variables to comply with field naming conventions in the code styles

juliawong committed:
Coordinates are now successfully being parsed in. Added a sample file

juliawong committed:
Styles are now held in a hashmap in the Document class. Minor refactoring of code.

Commit:
Removed getter methods in properties classes

juliawong committed:
Implemented defining the boundary of polygons and minor changes to support the styles hashmap

juliawong committed:
Fixed the Styles class to iterate through all styles until closing style tag is found

juliawong committed:
Modified Style class to store PolyStyle and LineStyle in the styleProperties method

juliawong committed:
Added methods in the Style class to allow the user to get a PolylineObject or PolygonObject

juliawong committed:
Added require methods to verify the expected XML structure

Commit:
Added features to the map

juliawong committed:
PolylineOptions and PolygonOptions are being built in this class as the styles are parsed. Added get methods for both.

juliawong committed:
Added a second constructor to directly read from a raw resource file

juliawong committed:
Reformatted all code in the KML directory

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Renamed and added public methods that will be supported

juliawong committed:
Performed a reformat of the code. Added and redid comments for classes and methods.

juliawong committed:
Renamed a file for demoing GeoJSON

Commit:
Implemented toggle feature

juliawong committed:
Implemented functions to add all elements in an ArrayList to mGeoJsonMapObjects

juliawong committed:
Changed color to be parsed from the GeoJSON file as a string instead of as a hexadecimal.

Commit:
Better color parsing, cleaned up code

Commit:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Supported nested GeometryCollection in GeometryCollections. Refactored code. Implemented method to remove overlay.

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added toString methods for all Properties classes

juliawong committed:
Included further support for GeometryCollections

juliawong committed:
Reformat comments

juliawong committed:
Allowed properties to be applied to GeometryCollections

juliawong committed:
Removed TODO

juliawong committed:
Added functions to show, hide and invert the visibility of the data on the map

juliawong committed:
Corrected PolygonProperties visibility

juliawong committed:
Implemented toggling of the overlay

lchans committed:
Fixed issue with method not assigning hole to Polygon

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added buttons to the demo

juliawong committed:
Cleaned up demo and renamed

juliawong committed:
Added support for Marker options flat and info window anchor

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Renamed public methods to comply with design doc

juliawong committed:
Prevent duplicate layers of the same data being added

juliawong committed:
Added an Add button to the GeoJSON demo

juliawong committed:
Reformatted code and updated comments

juliawong committed:
Added an add button and fixed variable name

juliawong committed:
Renamed class name to comply with style guidelines. Added javadoc for @throws

juliawong committed:
Modified to adapt to the renaming of the geoJsonImport class

juliawong committed:
Removed old files which have been renamed

juliawong committed:
Made methods static and made some minor modifications to comply with style guidelines.

juliawong committed:
Removed invertVisibility method

juliawong committed:
Removed toggleVisibility, showAllGeoJsonData and hideAllGeoJsonData methods. Added geoJsonDataVisibility method to replace removed method. Updated demo.

juliawong committed:
Removed a statement that was always true

juliawong committed:
Added and updated tests

juliawong committed:
Added folder and files for new GeoJSON API

juliawong committed:
Added methods for collection class

juliawong committed:
Ran reformatter

juliawong committed:
Added methods for feature class

juliawong committed:
Added geometry class

lchans committed:
Made geometry object implement the same interface

juliawong committed:
Added methods to all Geometry objects

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Moved files into correct directory

juliawong committed:
Added toString methods, constructors and a sample geometry method for debugging

juliawong committed:
Added Style classes

juliawong committed:
Added Options objects to style classes

juliawong committed:
Renamed type variables

juliawong committed:
Implemented functions in Collection and reformatted code

juliawong committed:
Added ability to read raw resource files

juliawong committed:
Added parser with ability to parse FeatureCollections and Features

juliawong committed:
Added parsing by a single geometry

juliawong committed:
Updated style classes with regex type and fixed minor bugs

juliawong committed:
Refactored to allow for id to be only set once

juliawong committed:
Added JavaDoc and reformatted code

juliawong committed:
Added Javadoc to GeoJsonParser

juliawong committed:
Added function to parse GeoJSON and toString

juliawong committed:
Added toString method

juliawong committed:
Removed old GeoJson utility library

juliawong committed:
Updated demo for new GeoJson utility

juliawong committed:
Added toString methods

juliawong committed:
:tada: Support for GeometryCollection. Now parses Geometry, Feature and FeatureCollections!"

lchans committed:
Implemented multigeometry

lchans committed:
Implemented multigeometry

juliawong committed:
:earth_asia: Added more demo classes. Fixed markers in Spain file. Updated demo file.

juliawong committed:
Implemented iterators for Collection and Feature classes

juliawong committed:
Removed old tests. Added skeleton for new tests.

juliawong committed:
Added the assigning of default styles in the GeoJsonParser

juliawong committed:
Refactored Style classes and added Javadoc

juliawong committed:
Added Javadoc summary to all files in GeoJsonLayer

juliawong committed:
Added constructors to initialise the Options objects

juliawong committed:
Added exceptions

juliawong committed:
Added testing for types and coordinates

lchans committed:
Added JavaDoc for existing classes

juliawong committed:
Implemented ability to add GeoJSON features and geometries onto the map

juliawong committed:
Added testing of ID

lchans committed:
Updating comments

juliawong committed:
Added new KML file with markers

juliawong committed:
Renamed demo file

juliawong committed:
Performed minor refactoring on KML files. Allowed Style class to create Options objects.

juliawong committed:
Minor modifications to demo layouts. New demo file added.

juliawong committed:
Refactored code. Started implementation of IconStyle

lchans committed:
GeoJson adds GeometryCollections to the map, also modified some API features

juliawong committed:
Implemented IconStyle. Updated demo.

juliawong committed:
Added a MultiGeometry class. Reallowed for adding to map.

lchans committed:
Objects are redrawn when styles are set

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Renamed files to be prefixed with Kml. Added more support for recursive MultiGeometries through creating a Parser class to separate the logic and handle the processing of the data.

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added more kml files for demos

juliawong committed:
:cookie: Points, LineStrings and Polygons are successfully being drawn onto the map

lchans committed:
Geometry collections can be added and removed from the map - Nested geometry collections are added to the features hashmap as an arraylist

juliawong committed:
:cow2: MultiGeometry adding to the map

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added test skeleton files

lchans committed:
Added tests for features

juliawong committed:
Added support for the normal style tag in StyleMap

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added public methods remove, set z index and get placemarks

juliawong committed:
Fixed minor syntax error and removed old code

lchans committed:
Added additional tests for GeoJsonParser

juliawong committed:
Modified KmlLayer to download icons once and add to many markers

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Minor refactor of KmlLayer to utilise LruCache

juliawong committed:
Refactored code to not store Marker objects twice, uses LruCache more effectively

juliawong committed:
Reformatted code, added TODO

juliawong committed:
Added exceptions for null coordinates. Renamed KmlGeometry method

juliawong committed:
Added tests for getType

juliawong committed:
Updated demo files and grade files

lchans committed:
Added comments

juliawong committed:
Changed GeoJsonGeometry to an interface

juliawong committed:
Minor syntax correction

lchans committed:
Seperated KMLParser into KMLStyleParser and KMLPlacemarkParser

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Refactored code, ran reformatter and added some JavaDoc

juliawong committed:
Renamed tests

juliawong committed:
Added support for bbox in FeatureCollection

juliawong committed:
Implemented addFeature, removeFeature methods

juliawong committed:
Refactored rendering of GeometryCollection

lchans committed:
Implemented Folders, however need to also create a Document class so that Folder items can inherit properties available

juliawong committed:
:balloon: Added and redid Javadoc. Minor refactor of code.

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Modified methods to add and remove data to allow developers to call add multiple times. Removed style interface.

juliawong committed:
Prevented feature redrawing if different type style was set

juliawong committed:
Features are redrawn when styles are updated

lchans committed:
Can handle nested folders and overriding styles in KML

juliawong committed:
Added tests for style classes

juliawong committed:
Minor clean ups

lchans committed:
Can fetch a url for a marker in folders

juliawong committed:
Caught exceptions and logging when GeoJSON could not be parsed.

juliawong committed:
Reformatting tests and minor clean ups

lchans committed:
Implemented API methods for Folders

juliawong committed:
Added more tests for null coordinates

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

lchans committed:
Implement public API methods for KMLLayer

juliawong committed:
KmlGeometry is now typed

Updated all references to getKmlGeometryCoordinates to add types and
remove casts.

lchans committed:
Added comments to code

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonFeature.java:
```diff
@@ -0,0 +1,214 @@
+package com.google.maps.android.geoJsonLayer;
```
no capitals in package name (http://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html). I might also just call it geojson instead of geojsonlayer

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonFeature.java:
```diff
@@ -0,0 +1,214 @@
+
+    private final String mId;
+
+    private final ArrayList<LatLng> mBoundingBox;
```
why not LatLngBounds here?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonFeature.java:
```diff
@@ -0,0 +1,214 @@
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("Feature{");
+        sb.append("\n bounding box=").append(mBoundingBox);
```
commas?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonFeature.java:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    @Override
+    public String toString() {
```
Is this support to return the JSON representation?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @param stream Character input stream representing the GeoJSON file
+     * @return JSONObject representing the GeoJSON file
+     * @throws java.io.IOException    if the file cannot be opened for read
```
no need for java.io

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+     * @param stream Character input stream representing the GeoJSON file
+     * @return JSONObject representing the GeoJSON file
+     * @throws java.io.IOException    if the file cannot be opened for read
+     * @throws org.json.JSONException if the JSON file has poor structure
```
ditto

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return iterator of GeoJsonFeature elements
+     */
+    public java.util.Iterator<GeoJsonFeature> getFeatures() {
```
returning Iterable might be more useful than returning Iterator because then developers can use it in for-each loops easily

juliawong replied:
Will return the keySet instead which is of type Set<GeoJsonFeature>

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+     */
+    public void addFeature(GeoJsonFeature feature) {
+        mFeatures.put(feature, null);
+        feature.addObserver(mRenderer);
```
Should this be done in mRenderer?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+        if (mFeatures.containsKey(feature)) {
+            mRenderer.removeFeature(feature);
+            mFeatures.remove(feature);
+            feature.deleteObserver(mRenderer);
```
responsibility of Renderer?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+        mFeatures.put(feature, null);
+        feature.addObserver(mRenderer);
+        mRenderer.addFeature(feature);
+    }
```
should the default styles be applied to this feature?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonStyle.java:
```diff
@@ -0,0 +1,14 @@
+ */
+public interface GeoJsonStyle {
+
+    public String getGeometryType();
```
comment this.

Instead of a regex, I would just use a String[]. Or even a Class<? extends GeoJsonGeometry>[] to give you some nice type safety.

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonMultiLineString.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final static String GEOMETRY_TYPE = "MultiLineString";
+
+    private ArrayList<GeoJsonLineString> mGeoJsonLineStrings;
```
final

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonMultiPoint.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final static String GEOMETRY_TYPE = "MultiPoint";
+
+    private ArrayList<GeoJsonPoint> mGeoJsonPoints;
```
final

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+ * array of
+ * GeoJsonFeature objects parsed from the GeoJSON file.
+ */
+public class GeoJsonParser {
```
package private instead of public? if so, do this elsewhere in this class as well

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+     * @param geoJsonFeature feature to parse
+     * @return GeoJsonFeature object
+     */
+    private GeoJsonFeature parseFeature(JSONObject geoJsonFeature) {
```
static

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+     * second element is the highest values
+     * @throws JSONException if the bounding box could not be parsed
+     */
+    private ArrayList<LatLng> parseBoundingBox(JSONArray coordinates) throws JSONException {
```
static

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+     * @param geoJsonGeometry geometry object to parse
+     * @return GeoJsonGeometry object
+     */
+    private GeoJsonGeometry parseGeometry(JSONObject geoJsonGeometry) throws JSONException {
```
static

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+     * @return hashmap containing property values
+     * @throws JSONException if the properties could not be parsed
+     */
+    private HashMap<String, String> parseProperties(JSONObject properties) throws JSONException {
```
static

and everywhere else in this class

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonParser.java:
```diff
@@ -0,0 +1,464 @@
+            return createPolygon(geometryArray);
+        } else if (geometryType.equals("MultiPolygon")) {
+            return createMultiPolygon(geometryArray);
+        } else if (geometryType.equals(GEOMETRY_COLLECTION)) {
```
why is GEOMETRY_COLLECTION a constant but the rest aren't?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonLayer.java:
```diff
@@ -0,0 +1,293 @@
+ */
+public class GeoJsonLayer {
+
+    private final HashMap<GeoJsonFeature, Object> mFeatures;
```
I think this should just be a List (or Set) of GeoJsonFeatures.  I don't think this class should know about the rendered objects.  That Map should be internal to the Renderer.

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonRenderer.java:
```diff
@@ -0,0 +1,308 @@
+     * @param features hashmap of features to store
+     * @param map      map to place GeoJsonFeature objects on
+     */
+    public GeoJsonRenderer(HashMap<GeoJsonFeature, Object> features, GoogleMap map) {
```
make all methods in this class package-private as well (except for those for which this is not possible like those that override Observer methods)

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonRenderer.java:
```diff
@@ -0,0 +1,308 @@
+     *
+     * @param feature feature to redraw onto the map
+     */
+    private void redrawCollectionToMap(GeoJsonFeature feature) {
```
why is this called redrawCollectionToMap - it only acts on a specific feature?

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonRenderer.java:
```diff
@@ -0,0 +1,308 @@
+     */
+    public void removeFeature(GeoJsonFeature feature) {
+        if (feature.hasGeometry()) {
+            removeFromMap(mFeatures.get(feature));
```
this method should also update the HashMap.  I actually think that you should do that in the removeFromMap method but you might have to change the logic slightly to be less recursive.

amorris13 commented on library/src/com/google/maps/android/geoJsonLayer/GeoJsonRenderer.java:
```diff
@@ -0,0 +1,308 @@
+     * @param feature feature to redraw onto the map
+     */
+    private void redrawCollectionToMap(GeoJsonFeature feature) {
+        removeFromMap(mFeatures.get(feature));
```
is there a way to do this without removing and adding to the map?

juliawong replied:
This is a future improvement. We'll put this in our design doc.

juliawong committed:
BoundingBox stored as a LatLngBounds object

juliawong committed:
Renamed package to geojson

lchans committed:
Made Documents a KmlContainer (Can now have nested Documents, however reference doesn't provide us with an indication of whether we can)

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Refactor to store map objects in GeoJsonRenderer and make methods static.

juliawong committed:
Made geometry arrays and coordinates final

Affects classes which implement GeoJsonGeometry

Added static final variables for geometries

juliawong committed:
Added constants for geometries

juliawong committed:
Minor refactor of GeoJsonRenderer

juliawong committed:
Renamed the GeoJsonMultiLineStringTest and added tests/

juliawong committed:
Renamed GeoJsonMultiPolygonTest and added tests

lchans committed:
Started on implementation for GroundOverlay

lchans committed:
Merge conflict

juliawong committed:
GeoJsonParser and GeoJsonRenderer are now package protected. Ran reformatter on geojson folder.

juliawong committed:
Fixed tests failing due to IBitmapDescriptorFactory

lchans committed:
Cleaned up comments

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Added more GeoJsonFeature tests

juliawong committed:
Removed TODO

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

markmcd commented on demo/res/raw/nestedfolders.kml~:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file looks like it shouldn't be here (duplicate of nestedfolders.kml)

juliawong committed:
Modified exception throwing in the parser. Added tests.

juliawong committed:
getFeatures returns an Iterable

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonFeature.java:
```diff
@@ -0,0 +1,211 @@
+     * @param GeoJsonGeometry type of geometry to assign to the feature
+     * @param id              id to refer to the feature by
+     * @param properties      hashmap of data containing properties related to the feature
+     * @param boundingBox     array defining the bounding box of the feature
```
Any reason why the bounding box is user-provided, rather than being calculated from the geometry?

juliawong replied:
We didn't want to calculate it because it's outside the scope of the project. It also might be tricky with polygons as well.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonFeature.java:
```diff
@@ -0,0 +1,211 @@
+     * the FeatureCollection did not have a bounding box or if the GeoJSON file did not contain a
+     * FeatureCollection then null will be returned.
+     *
+     * @return LatLngBounds containing bounding box of FeatureCollection, null if no bounding box
```
These docs mention FeatureCollection, I think it's meant to be Feature.

juliawong replied:
FeatureCollection is correct. FeatureCollections can have a bounding box but so can features as well. We're representing the layer as the feature collection as we don't have a class specifically for that.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonFeature.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Gets the GeoJsonLineStringStyle of the feature
```
Doc comments in this format don't provide any more information than is already available from the method signature. Consider something like "Returns the style used when rendering LineStrings".

juliawong replied:
Done. I've left the "@return" and "@param" as "GeoJsonPointStyle object". Does this also need to be changed?

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonGeometry.java:
```diff
@@ -0,0 +1,14 @@
+public interface GeoJsonGeometry {
+
+    /**
+     * Gets the type of geometry
```
Can you include some examples here (or in the implementation classes): "e.g. "LineString", "MultiPoint"" - or mention that it will conform to the GeoJson 'type' specification.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonGeometryCollection.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * Creates a new GeoJsonGeometryCollection object
+     *
+     * @param geometries array of GeoJsonGeometry objects to add to the GeoJsonGeometryCollection
```
nit: 'array of ...' does not match the method signature, which requires an ArrayList. You also say "to add to the ...", which implies that it's additive, which it is not.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonGeometryCollection.java:
```diff
@@ -0,0 +1,53 @@
+     * @param geometries array of GeoJsonGeometry objects to add to the GeoJsonGeometryCollection
+     */
+    public GeoJsonGeometryCollection(
+            ArrayList<GeoJsonGeometry> geometries) {
```
Does this need to be an ArrayList? Or is List or Collection enough?

juliawong replied:
Done. Changed to List. Can't use collection as we use the get method

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonGeometryCollection.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    /**
+     * Gets the type of geometry
```
Since this is the same as the parent, I think you can omit the javadoc completely. Or use @inheritdoc or something like that.

juliawong replied:
Done. Using /*\* {@inheritDoc} */

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLayer.java:
```diff
@@ -0,0 +1,251 @@
+    /**
+     * Creates a new GeoJsonLayer object
+     *
+     * @param map        GoogleMap object
```
"GoogleMap object" is implied by the method signature. Same for anywhere you've used "<Type> object" in the doc. For this one, you can say something like "The map where the layer is to be rendered."

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLayer.java:
```diff
@@ -0,0 +1,251 @@
+
+
+/**
+ * A class that allows the developer import GeoJSON data, style it and interact with the imported
```
I think this class is complex enough that you should include a {@code} block showing a code sample of how to use it in the JavaDoc here.

juliawong replied:
Done. Similar to code in presentation.

juliawong committed:
Reformat and minor refactor of code. Changed all iterators to iterables. Added some more Javadoc.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLayer.java:
```diff
@@ -0,0 +1,251 @@
+     * Removes all of the stored GeoJsonFeature objects from the map and clears the mFeatures
+     * hashmap
+     */
+    public void removeGeoJsonLayer() {
```
I think this could be named better. 'remove layer', to me at least, implies that the layer remains intact but is no longer attached to the map, whereas this (I think) actually empties out the layer completely, clearing out all data. Perhaps something like 'clear()' or 'reset()'.

juliawong replied:
Good point, renaming to clearLayer()

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLayer.java:
```diff
@@ -0,0 +1,251 @@
+     *
+     * @param geoJsonLineStringStyle to set as default style
+     */
+    public void setDefaultLineStringStyle(GeoJsonLineStringStyle geoJsonLineStringStyle) {
```
Can you elaborate in the docs as to what the relationship is between the default styles and the styles attached to the features? i.e. will the default be rendered iff the feature has no style? or will the default be rendered first, then updated with the local style? or something else?

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLineString.java:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @param coordinates array of coordinates of GeoJsonLineString to store
+     */
+    public GeoJsonLineString(ArrayList<LatLng> coordinates) {
```
As above, does this need to be an ArrayList? If List or Collection would do, use them instead.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLineStringStyle.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    /**
+     * Sets the color of the GeoJsonLineString
```
What does the integer mean? Is it an RGB formatted like 0xRRGGBB? RGBA? Is it a resource ID?

juliawong replied:
Done. 32-bit ARGB color

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLineStringStyle.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    /**
+     * Sets the width of the GeoJsonLineString
```
What are the units here?

juliawong replied:
Done. Screen pixels.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonLineStringStyle.java:
```diff
@@ -0,0 +1,168 @@
+     *
+     * @return new PolylineOptions object
+     */
+    public PolylineOptions getPolylineOptions() {
```
Convention for something like this is 'toPolylineOptions()', as 'get..' implies a local property of this object. Whereas what's happening here is a conversion.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonPoint.java:
```diff
@@ -0,0 +1,51 @@
+import com.google.android.gms.maps.model.LatLng;
+
+/**
+ * A GeoJsonPoint geometry contains a single {@link com.google.android.gms.maps.model.LatLng}.
```
Slight nit: make a note that this class does not support altitude or any further values. Technically GeoJson allows a point or position to be an unbounded list of values representing "lat, lng, altitude, undefined, ..."

juliawong replied:
Adding this to the GeoJsonGeometry interface as it applied to all geometry objects.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonPointStyle.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    /**
+     * Sets the alpha of the GeoJsonPoint
```
What are the units here? 0..1 for opaque to transparent? -MAX_FLOAT to +MAX_FLOAT? -1 to 1?

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonPointStyle.java:
```diff
@@ -0,0 +1,297 @@
+     * @return type of geometries this style can be applied to
+     */
+    @Override
+    public String getGeometryType() {
```
Is this expected to be used outside of this library? If not, drop the privacy to package-private or something appropriate. Also, you can use Pattern instead of String if the value truly represents a regex pattern.

juliawong replied:
Yes, we want to allow developers to check the type of object.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonPointStyle.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    /**
+     * Sets the Anchor U and V coordinates of the GeoJsonPoint
```
Make a note that U & V represent the same U & V values used in MarkerOptions.

juliawong replied:
Done.

markmcd commented on library/src/com/google/maps/android/geojson/GeoJsonPointStyle.java:
```diff
@@ -0,0 +1,297 @@
+
+
+    /**
+     * Sets the rotation of the GeoJsonPoint
```
Units?

juliawong replied:
Done.

markmcd commented:
Overall - lookin' pretty good. There are a few places I've made comments that apply in multiple places but I've not made multiple comments (e.g. units, arraylists).

juliawong committed:
Fixed error preventing inner containers from being created.

juliawong committed:
Implemented parsing of extended data in features

juliawong committed:
Implemented parsing of extended data in containers

juliawong committed:
Implemented parsing and adding GroundOverlay onto the map

juliawong committed:
Added toString methods

lchans committed:
Throws an error if the image could not be retrieved

lchans committed:
Throwing error if no text found in BalloonStyle

lchans committed:
Added Polygon tests, fixed error where not existent inner boundaries were being attempted to add

lchans committed:
Added tests for KML geometry objects and basic KML style functions

lchans committed:
Added more tests for styles

lchans committed:
Added tests for container parser, changed visibility of containers to be visible publicly in package only

juliawong committed:
Support for inline styles in placemarks

juliawong committed:
Updated demo files

juliawong committed:
Removed KmlContainerInterface and updated references to it

lchans committed:
Added tests for KmlContainerParser, focusing mainly on nested containers

juliawong committed:
Inline styles stored for containers. Separated from shared styles.

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
One LruCache being used for marker icons and ground overlays

lchans committed:
Added additional tests for nested folders

juliawong committed:
Removed inline styles for containers, not supported in KML

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

lchans committed:
Added seperate method for creating a latlngbound for better readability

lchans committed:
Changed FeatureParser from if to else if statements

lchans committed:
Added testing for feature, now fetches test files from a raw resource rather than as a string

juliawong committed:
Refactored part of KmlLayer into KmlRenderer

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

lchans committed:
Added tests for multigeometry and added a hasProperties method

lchans committed:
Fixed up a few javadoc commments and also deleted code so that it is cleaner

lchans committed:
Added tests for multiple features in a container

juliawong committed:
Implemented inline styles

lchans committed:
Added extended data test

lchans committed:
Color mode working for markers

lchans committed:
Seperated methods, now has inlinestyle methods

lchans committed:
Refractored variable names

lchans committed:
Added style map tests

juliawong committed:
Support for inline icon URLs and scale

lchans committed:
Refractored variable names

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

lchans committed:
Added more ground overlay tests

lchans committed:
Added testing for cdata entities

lchans committed:
Changed to nextToken

lchans committed:
Changed implemented of createGroundOverlay to be seperate create and get methods, also cleaned up javadoc comments in KmlContainerParser

lchans committed:
Ground Overlay stores color

juliawong committed:
Removal of GroundOverlays, parse objects once, download images once

lchans committed:
Added more javadoc

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

lchans committed:
Point displays an info window even if balloonstyle is not set (it displays when description is set), changed implementation so that an info window displays in this case

juliawong committed:
Closed the stream

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Using additional hashmap to assign styles

lchans committed:
Changed javadoc comments

lchans committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Minor modification to allow for style maps to be updated only when reassigned

juliawong committed:
Merge branch 'new' of https://github.com/juliawong/android-maps-utils into new

juliawong committed:
Reformat parser
